driver-core/src/main/com/mongodb/internal/connection/AbstractProtocolExecutor.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.internal.connection;
+
+import com.mongodb.MongoOperationTimeoutException;
+import com.mongodb.MongoSocketException;
+import com.mongodb.internal.session.SessionContext;
+
+/**
+ * <p>This class is not part of the public API and may be removed or changed at any time</p>
+ */
+public abstract class AbstractProtocolExecutor implements ProtocolExecutor {
```

Comment:
Does this abstract class need to exist? Can these helper methods just be added to `ProtocolExecutor`?

Reply:
Technically, the methods could be moved to the interface since they are stateless, but this would make them public. I've opted to keep them `protected` within the abstract class to maintain better encapsulation and clearer boundaries in our codebase. If you feel that we should move them to the interface, I am open to doing that. Please let me know your thoughts.

I've also relocated the `isMongoSocketException `and `isOperationTimeoutFromSocketException` to `ExceptionUtils` for broader reuse.

Reply:
This is all `internal` code, it just adds an extra layer for helper style methods. Looks like this abstract class could be removed and `shouldMarkSessionDirty` be moved to a helper.

However, if you are keen to keep it protected then its ok with me.